amp-pwa/src/components/amp-document.js

```diff
@@ -0,0 +1,117 @@
+import React from 'react';
+
+/**
+ * Fetches the AMP document at a given `src` URL and renders it via Shadow DOM.
+ */
+export default class AMPDocument extends React.Component {
+  constructor(props) {
+    super(props);
+
+    // 'offline' is set to true if and when the document fetch fails.
+    this.state = {'offline': false};
+
+    /**
+     * `window.AMP` is set by the AMP runtime when it finishes loading.
+     * @private
+     */
+    this.ampReadyPromise_ = new Promise(resolve => {
+      (window.AMP = window.AMP || []).push(resolve);
+    });
+
+    /**
+     * Child element that the AMP document will be added as a shadow root to.
+     * @private
+     */
+    this.container_ = null;
+
+    /**
+     * XMLHTTPRequest that fetches the AMP document.
+     * @private
+     */
+    this.xhr_ = null;
+  }
+
+  componentDidMount() {
+    this.fetchAndAttachAMPDoc_(this.props.src);
+  }
+
+  componentWillUnmount() {
+    if (this.xhr_) {
+      this.xhr_.abort();
+      this.xhr_ = null;
+    }
+  }
+
+  componentWillReceiveProps(nextProps) {
+    this.fetchAndAttachAMPDoc_(nextProps.src);
+  }
+
+  render() {
+    if (this.state.offline) {
+      return (
+        <div>
+          <h2>Houston, we have a problem.</h2>
+          <p>Looks like we are offline&mdash;please check your Internet connection.</p>
+        </div>
+      );
+    } else {
+      return (<div className='amp-container' ref={ref => this.container_ = ref} />);
+    }
+  }
+
+  /**
+   * Fetches the AMP document at `url` and attaches it as a shadow root.
+   * @private
+   * @param {string} url
+   */
+  fetchAndAttachAMPDoc_(url) {
+    this.fetchDocument_(url).then(doc => {
+      return this.ampReadyPromise_.then(amp => {
+        amp.attachShadowDoc(this.container_, doc, url);
+      });
+    }).catch(error => {
+      this.setState({'offline': true});
+    });
+  }
+
+  /**
+   * Fetches and parses HTML at `url`.
+   * @private
+   * @param {string} url
+   * @return {Promise} If fetch succeeds, resolved with {Document}.
+   *         Otherwise, rejects with {string} error description.
+   */
+  fetchDocument_(url) {
+    return new Promise((resolve, reject) => {
+      const xhr = (this.xhr_ = new XMLHttpRequest());
```

Comment:
Looks like we also need to reset `this.xhr_ = null` in the either outcome of this promise.